File: Practice/atcoder/ABC/054/src/d2.rb
# ナップザック
N, A, B = gets.chomp.split.map(&:to_i)
@a_arr, @b_arr, @c_arr = [], [], []
N.times do
    a, b, c = gets.chomp.split.map(&:to_i)
    @a_arr << a; @b_arr << b; @c_arr << c
end
@memo = Array.new(N).map{Array.new(40*10+1).map{Array.new(40*10+1, -1)}}
def rec(i, a, b, c)
    # 見つかった場合は終了
    if a*b>0 && a*B == A*b
        return c
    end
    # 探索できなかったら終了
    if i==N
        return 10**10
    end
    # 既に探索済みならそれを返す
    if @memo[i][a][b] > 0
        return @memo[i][a][b]
    end
    ret = [rec(i+1, a+@a_arr[i], b+@b_arr[i], c+@c_arr[i]), rec(i+1, a, b, c)].min
    @memo[i][a][b] = ret
    return ret
end

ans = rec(0, 0, 0, 0)
puts ans==10**10 ? -1 : ans

File: Practice/atcoder/ABC/119/src/d2.py
# Nが小さいので全探索の構え -> なにをどうする？
# A,B,Cしかないので「li本目をA,B,Cどれの材料にするか」でやれそう -> 振り分けてから和をとって（+=10）それぞれの差を計算(+=1)
n, a, b, c = map(int, input().split())
arr = []
for _ in range(n):
    arr.append(int(input()))


def rec(i, ai, bi, ci):
    if i == n:
        if ai == 0 or bi == 0 or ci == 0:
            # 各最低でも1本以上必要だよね
            return 10 ** 10
        return abs(a - ai) + abs(b - bi) + abs(c - ci)
    else:
        ret = 10 ** 10
        # いま注目してるやつをa,b,cどれに属することにするか選ぶ
        if ai == 0:
            ret = min(ret, rec(i + 1, arr[i], bi, ci))
        elif ai > 0:
            ret = min(ret, rec(i + 1, ai + arr[i], bi, ci) + 10)
        if bi == 0:
            ret = min(ret, rec(i + 1, ai, arr[i], ci))
        elif bi > 0:
            ret = min(ret, rec(i + 1, ai, bi + arr[i], ci) + 10)
        if ci == 0:
            ret = min(ret, rec(i + 1, ai, bi, arr[i]))
        elif ci > 0:
            ret = min(ret, rec(i + 1, ai, bi, ci + arr[i]) + 10)
        # 今回は選ばない場合もある
        ret = min(ret, rec(i + 1, ai, bi, ci))
        return ret


print(rec(0, 0, 0, 0))
